qiskit/utils/backend_utils.py

```diff
@@ -34,6 +34,23 @@ def __init__(self) -> None:
 _PROVIDER_CHECK = ProviderCheck()
 
 
+def _get_backend_version(backend):
+    """Get the backend version int."""
+    backend_version = getattr(backend, "version", None)
+    if not isinstance(backend_version, int):
+        backend_version = 0
+    return backend_version
```

Comment:
The difference between `version` and `backend_version` is already a bit confusing. I mean `backend.version` and `backend.backend_version` both exist. But, here the class variable `backend.version` is called `backend_version` and this is repeated in the callers. I imagine these names have already been discussed. Maybe "revision" for what is now `backend.backend_version`? That's not very good, but what we have now might be worse.

Also, since it's confusing, I can imagine someone might make a mistake like this:
```python
 In [1]: from qiskit.test.mock.fake_backend_v2 import FakeBackendSimple  # or another user/defined backend

In [2]: fb = FakeBackendSimple()

In [3]: fb.version = "0.0.2"  # maybe this is done in __init__

In [4]: from qiskit.utils.backend_utils import _get_backend_version

In [5]: _get_backend_version(fb)
Out[5]: 0
```
Something like this would catch such a mistake:
```python
    if backend is None:
      backend _version = 0
    elif not isinstance(backend_version, int):
        raise TypeError("Expecting an integer for backend_version ...")
```

Reply:
Yeah I can see this perspective I can try to pick a different name for the local variable to try and address this concern here, but changing names in the backend interface now would be a breaking change since it's in the wild already.

The thing which makes it clearer to me is knowing that `backend.version` (for BackendV1, V2, and any new versions moving forward) is actually a class attribute defined in the abstract interface which defines the interface version. It's actually the only consistent attribute between all versions of the versioned backend interface: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/providers/backend.py#L40. It's not designed to be user settable (but as you said it is in practice because python, but doing that would be a bad idea because it breaks the interface). While `backend.backend_version` is optional metadata a backend can specify (although in practice only ibm backends really use it) to specify a custom version string. It's set via `super().__init__()` in implementations of `BackendV2`. That being said, personally I've never actually used `backend.backend_version` for anything (both as a backend author and as a user interacting with backends as part of executing circuits or other jobs) since it's opaque to me as a user what it means in practice.

Reply:
I changed the local variable names in: https://github.com/Qiskit/qiskit-terra/pull/7563/commits/bfd981a261639e2545d2f5524c369e005e8121f8 to be a bit more descriptive

Reply:
I think the best way to probably handle this is to change the `Backend` class to be:

```python
class Backend:
    _version = 0
    
    @property
    def version(self):
        return self._version
```
then in each interface version we change `version =` to `_version =`. This should be backwards compatible (but we might want to save it for 0.20.0 just to be safe) as accessing `backend.version` should work identically, but this does mean that attribute can't be changed directly. A user could still do `backend._version = "IamAversion"` to break things with that, but it seems less likely for that to happen by mistake. I can open a separate PR doing that later

Reply:
Yes. I thought about this one as well. It definitely puts up more of a barrier. I didn't mention it because `BackendVx` do not inherit from `Backend`. This would have to be changed, right ?